crates/astria-sequencer-relayer/src/finalization_pipeline/mod.rs

```diff
@@ -0,0 +1,194 @@
+use std::{
+    collections::HashMap,
+    mem,
+};
+
+use astria_sequencer_types::SequencerBlockData;
+use tendermint::Hash;
+
+pub(crate) mod head_candidate;
+pub(crate) mod pipeline_item;
+
+pub(crate) use head_candidate::HeadCandidate;
+pub(crate) use pipeline_item::PipelineItem;
+
+// head of the sequencer chain as observed on cometbft.
+#[derive(Default)]
+pub(crate) struct Head {
+    block: PipelineItem,
+}
+
+// pipeline can handle forks 1 height deep
+#[derive(Default)]
+pub(crate) struct FinalizationPipeline {
+    pub(crate) chain_head: Head,
+    pending: HashMap<Hash, PipelineItem>,
+    validator_finalized: Vec<SequencerBlockData>,
+}
+
+impl FinalizationPipeline {
+    pub(crate) fn submit(&mut self, new_block: HeadCandidate) {
+        let new_block: PipelineItem = new_block.into();
+
+        match new_block.parent_block_hash() {
+            Some(new_block_parent) => {
+                debug_assert!(new_block.height() > self.chain_head.block.height());
+
+                let steps = new_block.height() - self.chain_head.block.height();
```

Comment:
if the new_block is below or at the chain head height, this does nothing - there's a chance we receive two blocks with the same height (in the case where the first one fails to finalize for whatever reason) that case should be handled here

Reply:
the `chain_head.height()` here refers to the head of the canonical chain, i.e. the head is finalised. non-finalised blocks are head candidates.

Reply:
is it better to use `soft_block` for head and `head` for head candidates like in specs for conductor? https://github.com/emhane/astria/blob/da-finalized-blocks/specs/conductor.md